MojoExecutionException: Invalid image creation time: 2020-04-21T13:22:10.836777828-07:00

<!--
Before filing an issue, please reproduce with the latest version of Jib.

If you are encountering errors pulling or pushing to remote registries,
please check our Frequently Asked Questions before filing an issue:

    https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#registry-errors
-->

**Environment**:

- *Jib version:* 2.2.0
- *Build tool:* <!-- Maven/Gradle, including version --> docker image `maven:3-jdk-11`
- *OS:* docker image `maven:3-jdk-11`

**Description of the issue**:

The JIB build passes with builder image `maven:3-jdk-14` in GitLab-CI, but fails on `maven:3-jdk-11`. The error message is weird and I cannot reproduce it if JDK > 11.

In my local environment the build passes.
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/Cellar/maven/3.6.3_1/libexec
Java version: 13.0.2, vendor: N/A, runtime: /usr/local/Cellar/openjdk/13.0.2+8_2/libexec/openjdk.jdk/Contents/Home
Default locale: en_CN, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.4", arch: "x86_64", family: "mac"
```

**Expected behavior**:

Build on `maven:3-jdk-11`.

**Steps to reproduce**:
<!-- Please provide a minimal and precise series of steps -->

  1. Build command is:
```shell
mvn --batch-mode --errors --fail-at-end --show-version \
        compile jib:build \
        -pl ${SUB_MODULE_NAME} \
        -Djib.console=plain \
        -Djib.httpTimeout=60000 \
        -Djib.baseImageCache=$CI_PROJECT_DIR/.jib/base \
        -Djib.applicationCache=$CI_PROJECT_DIR/.jib/app \
        -Djib.to.auth.username=$CI_REGISTRY_USER \
        -Djib.to.auth.password=$CI_REGISTRY_PASSWORD
        -Djib.from.image=XXXXX \
        -Djib.to.image=XXXXXXXX
```

**`jib-maven-plugin` Configuration**: <!-- Delete this section if not used -->
```xml
...
      <plugin>
        <artifactId>jib-maven-plugin</artifactId>
        <groupId>com.google.cloud.tools</groupId>
        <version>2.2.0</version>
        <configuration>
          <container>
            <format>OCI</format>
            <jvmFlags>
              <jvmFlag>-server</jvmFlag>
              <jvmFlag>-Dio.netty.tryReflectionSetAccessible=false</jvmFlag>
            </jvmFlags>
            <user>1001:1001</user>
            <volumes>
              <volume>/data</volume>
            </volumes>
            <workingDirectory>/app</workingDirectory>
          </container>
          <from>
            <image>XXXXX</image>
          </from>
          <to>
            <image>XXXXXXXXX</image>
          </to>
        </configuration>
      </plugin>
...
```

**Log output**:

```
23488 [INFO] --- jib-maven-plugin:2.2.0:build (default-cli) @ stream ---
 24473 [INFO] 
 24476 [INFO] Containerizing application to XXXXXXXX...
 24497 [WARNING] Base image 'XXXXXXXX' does not use a specific image digest - build may not be reproducible
 24627 [INFO] Using credentials from <to><auth> for registry.gitlab.com/XXXXXXXX
 24634 [INFO] Getting manifest for base image XXXXXXXX...
 24672 [INFO] Building dependencies layer...
 24682 [INFO] Building snapshot dependencies layer...
 24689 [INFO] Building resources layer...
 24698 [INFO] Building classes layer...
 26326 [INFO] The base image requires auth. Trying again for XXXXXXXX...
 26894 [INFO] Using base image with digest: sha256:c181b1141de2f559444b6e3867a611d3c77ca9c02c9893808353341e1740f34b
 27244 [INFO] 
 27246 [INFO] ------------------------------------------------------------------------
 27249 [INFO] BUILD FAILURE
 27249 [INFO] ------------------------------------------------------------------------
 27255 [INFO] Total time:  25.008 s
 27259 [INFO] Finished at: 2020-04-27T03:57:09Z
 27265 [INFO] ------------------------------------------------------------------------
 27271 [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.2.0:build (default-cli) on project stream: Invalid image creation time: 2020-04-21T13:22:10.836777828-07:00: Text '2020-04-21T13:22:10.836777828-07:00' could not be parsed at index 29 -> [Help 1]
 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.2.0:build (default-cli) on project stream: Invalid image creation time: 2020-04-21T13:22:10.836777828-07:00
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
     at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
     at java.lang.reflect.Method.invoke (Method.java:566)
     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
 Caused by: org.apache.maven.plugin.MojoExecutionException: Invalid image creation time: 2020-04-21T13:22:10.836777828-07:00
     at com.google.cloud.tools.jib.maven.BuildImageMojo.execute (BuildImageMojo.java:162)
     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
     at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
     at java.lang.reflect.Method.invoke (Method.java:566)
     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
 Caused by: com.google.cloud.tools.jib.image.json.BadContainerConfigurationFormatException: Invalid image creation time: 2020-04-21T13:22:10.836777828-07:00
     at com.google.cloud.tools.jib.image.json.JsonToImageTranslator.configureBuilderWithContainerConfiguration (JsonToImageTranslator.java:151)
     at com.google.cloud.tools.jib.image.json.JsonToImageTranslator.toImage (JsonToImageTranslator.java:136)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.pullBaseImage (PullBaseImageStep.java:253)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call (PullBaseImageStep.java:160)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call (PullBaseImageStep.java:56)
     at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
     at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
     at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
     at java.lang.Thread.run (Thread.java:834)
 Caused by: java.time.format.DateTimeParseException: Text '2020-04-21T13:22:10.836777828-07:00' could not be parsed at index 29
     at java.time.format.DateTimeFormatter.parseResolved0 (DateTimeFormatter.java:2046)
     at java.time.format.DateTimeFormatter.parse (DateTimeFormatter.java:1948)
     at java.time.Instant.parse (Instant.java:395)
     at com.google.cloud.tools.jib.image.json.JsonToImageTranslator.configureBuilderWithContainerConfiguration (JsonToImageTranslator.java:148)
     at com.google.cloud.tools.jib.image.json.JsonToImageTranslator.toImage (JsonToImageTranslator.java:136)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.pullBaseImage (PullBaseImageStep.java:253)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call (PullBaseImageStep.java:160)
     at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call (PullBaseImageStep.java:56)
     at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
     at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
     at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
     at java.lang.Thread.run (Thread.java:834)
 27370 [ERROR] 
 27370 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
 27370 [ERROR] 
 27370 [ERROR] For more information about the errors and possible solutions, please read the following articles:
 27370 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Additional Information**: <!-- Any additional information that may be helpful -->


<!-- Thanks for contributing! -->

--- a/jib-core/CHANGELOG.md
+++ b/jib-core/CHANGELOG.md
@@ -28,6 +28,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 - Fixes the problem not inheriting `USER` container configuration from a base image. ([#2421](https://github.com/GoogleContainerTools/jib/pull/2421))
+- Fixed an issue when using a base image whose image creation timestamp contains timezone offset. ([#2428](https://github.com/GoogleContainerTools/jib/issues/2428))
 
 ## 0.13.1
 

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/image/json/JsonToImageTranslator.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/image/json/JsonToImageTranslator.java
@@ -32,6 +32,7 @@ import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
 import java.time.Duration;
 import java.time.Instant;
+import java.time.format.DateTimeFormatter;
 import java.time.format.DateTimeParseException;
 import java.util.ArrayList;
 import java.util.List;
@@ -147,8 +148,16 @@ public class JsonToImageTranslator {
       try {
         imageBuilder.setCreated(Instant.parse(containerConfigurationTemplate.getCreated()));
       } catch (DateTimeParseException ex) {
-        throw new BadContainerConfigurationFormatException(
-            "Invalid image creation time: " + containerConfigurationTemplate.getCreated(), ex);
+        try {
+          // TODO: remove when using Java >= 12.
+          // See https://github.com/GoogleContainerTools/jib/issues/2428
+          imageBuilder.setCreated(
+              DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(
+                  containerConfigurationTemplate.getCreated(), Instant::from));
+        } catch (DateTimeParseException ignored) {
+          throw new BadContainerConfigurationFormatException(
+              "Invalid image creation time: " + containerConfigurationTemplate.getCreated(), ex);
+        }
       }
     }
 

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/image/json/JsonToImageTranslatorTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/image/json/JsonToImageTranslatorTest.java
@@ -86,6 +86,28 @@ public class JsonToImageTranslatorTest {
   }
 
   @Test
+  public void testToImage_canParseTimestampWithOffset()
+      throws IOException, LayerPropertyNotFoundException, URISyntaxException,
+          LayerCountMismatchException, BadContainerConfigurationFormatException {
+    Path containerConfigJson =
+        Paths.get(
+            getClass().getClassLoader().getResource("core/json/containerconfig.json").toURI());
+    ContainerConfigurationTemplate containerConfig =
+        JsonTemplateMapper.readJsonFromFile(
+            containerConfigJson, ContainerConfigurationTemplate.class);
+    containerConfig.setCreated("2020-04-21T13:22:10.836777828-07:00");
+
+    Path manifestJson =
+        Paths.get(getClass().getClassLoader().getResource("core/json/v22manifest.json").toURI());
+    V22ManifestTemplate manifest =
+        JsonTemplateMapper.readJsonFromFile(manifestJson, V22ManifestTemplate.class);
+
+    // Should not throw BadContainerConfigFormatException.
+    // https://github.com/GoogleContainerTools/jib/issues/2428
+    JsonToImageTranslator.toImage(manifest, containerConfig);
+  }
+
+  @Test
   public void testPortMapToList() throws BadContainerConfigurationFormatException {
     ImmutableSortedMap<String, Map<?, ?>> input =
         ImmutableSortedMap.of(

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -10,6 +10,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 - Fixes the problem not inheriting `USER` container configuration from a base image. ([#2421](https://github.com/GoogleContainerTools/jib/pull/2421))
+- Fixed an issue when using a base image whose image creation timestamp contains timezone offset. ([#2428](https://github.com/GoogleContainerTools/jib/issues/2428))
 
 ## 2.2.0
 

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -10,6 +10,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 - Fixes the problem not inheriting `USER` container configuration from a base image. ([#2421](https://github.com/GoogleContainerTools/jib/pull/2421))
+- Fixed an issue when using a base image whose image creation timestamp contains timezone offset. ([#2428](https://github.com/GoogleContainerTools/jib/issues/2428))
 
 ## 2.2.0
 

